Better Debug impl for io::Error.

clarcharr opened the pull request:
This PR includes the below changes:

1. The former impl wrapped the entire thing in `Error { repr: ... }` which was unhelpful; this has been removed.
2. The `Os` variant of `io::Error` included the code and message, but not the kind; this has been fixed.
3. The `Custom` variant of `io::Error` included a `Custom(Custom { ... })`, which is now just `Custom { ... }`.

Example of previous impl:

```rust
Error {
    repr: Custom(
        Custom {
            kind: InvalidData,
            error: Error {
                repr: Os {
                    code: 2,
                    message: "no such file or directory"
                }
            }
        }
    )
}
```

Example of new impl:

```rust
Custom {
    kind: InvalidData,
    error: Os {
        code: 2,
        kind: NotFound,
        message: "no such file or directory"
    }
}
```

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit f058dce has been approved by `dtolnay`

<!-- @bors r=dtolnay f058dceb9de5ed22c1168fb07c55316f383fded7 -->

clarcharr commented:
I forgot a stability attribute so you'll have to `r+` again.

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit 418d050 has been approved by `dtolnay`

<!-- @bors r=dtolnay 418d0501b0a3c6187815d25d53a1d1e96a8bf431 -->

frewsxcv commented:
@bors rollup

frewsxcv commented:
```
[01:09:58] failures:
[01:09:58]     io::error::test::test_debug_error
[01:09:58] 
[01:09:58] test result: FAILED. 788 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
```

@bors r-

ollie27 commented on src/libstd/io/error.rs:
```diff
@@ -62,12 +62,18 @@ pub type Result<T> = result::Result<T, Error>;
     repr: Repr,
 }
 
+#[stable(feature = "io_error_debug", since = "1.24.0")]
```
As  `Debug` has been implemented for `io::Error` since 1.0.0 `#[stable(feature = "rust1", since = "1.0.0")]` would make more sense here.

clarcharr replied:
Will do

clarcharr commented:
I fixed the test, changed the stability attribute, and also added a `kind` field to the OS error debug considering how that wasn't included but could be useful.

dtolnay commented:
Could you update your comment at the top of this thread with an accurate before and after? The after `Os` representation should include the new `kind` field and the `Custom` looks like it should be getting formatted as a tuple struct according to the code.

clarcharr commented:
@dtolnay done! I also modified the failing test with one that's closer to the one listed in the OP.

kennytm commented on src/libstd/io/error.rs:
```diff
@@ -563,13 +571,32 @@ mod test {
+            }
+        };
+        let expected = format!(
+            "Custom { \
```
Should be `{{` here.

```
[01:18:41] error: invalid format string: expected `'}'`, found `'k'`
[01:18:41]    --> /checkout/src/libstd/io/error.rs:590:13
[01:18:41]     |
[01:18:41] 590 | /             "Custom { \
[01:18:41] 591 | |                 kind: InvalidInput, \
[01:18:41] 592 | |                 error: Os {{ \
[01:18:41] 593 | |                     code: {:?}, \
[01:18:41] ...   |
[01:18:41] 596 | |                 }} \
[01:18:41] 597 | |             }}",
[01:18:41]     | |_______________^
```

clarcharr commented:
I'm fairly certain that I've fixed the build errors. Waiting on CI.

dtolnay commented:
Travis says:

```
error[E0308]: mismatched types
   --> /checkout/src/libstd/io/error.rs:582:32
    |
582 |               repr: Repr::Custom(Custom {
    |  ________________________________^
583 | |                 kind: ErrorKind::InvalidInput,
584 | |                 error: box Error {
585 | |                     repr: super::Repr::Os(code)
586 | |                 },
587 | |             })
    | |_____________^ expected struct `realstd::boxed::Box`, found struct `io::error::Custom`
    |
    = note: expected type `realstd::boxed::Box<io::error::Custom>`
               found type `io::error::Custom`
```

shepmaster commented:
Ping from triage, @clarcharr! Will you have time to address the build issues?

killercup commented:
Build issue seems to be addressed, CI is green, so I'm gonna go ahead and

@bors r=dtolnay

bors commented:
:pushpin: Commit 52e074e has been approved by `dtolnay`

<!-- @bors r=dtolnay 52e074e40ee3eaf04415b2044c93bcfddad52f35 -->
